control/trajectory_follower/src/vehicle_model/vehicle_model_bicycle_kinematics.cpp

```diff
@@ -51,19 +51,23 @@ void KinematicsBicycleModel::calculateDiscreteMatrix(
 
   a_d << 0.0, velocity, 0.0, 0.0, 0.0, velocity / m_wheelbase * cos_delta_r_squared_inv, 0.0, 0.0,
     -1.0 / m_steer_tau;
-  Eigen::MatrixXd I = Eigen::MatrixXd::Identity(m_dim_x, m_dim_x);
-  a_d = (I - dt * 0.5 * a_d).inverse() * (I + dt * 0.5 * a_d);  // bilinear discretization
 
   b_d << 0.0, 0.0, 1.0 / m_steer_tau;
-  b_d *= dt;
 
   c_d << 1.0, 0.0, 0.0, 0.0, 1.0, 0.0;
 
   w_d << 0.0,
     -velocity * m_curvature +
       velocity / m_wheelbase * (tan(delta_r) - delta_r * cos_delta_r_squared_inv),
     0.0;
-  w_d *= dt;
+
+  // bilinear discretization
+  Eigen::MatrixXd I = Eigen::MatrixXd::Identity(m_dim_x, m_dim_x);
+  const Eigen::MatrixXd inv = (I - dt * 0.5 * a_d).inverse();
+  a_d = inv * (I + dt * 0.5 * a_d);
+  c_d = c_d * inv;
+  b_d = inv * b_d * dt;
+  w_d = inv * w_d * dt;
```

Comment:
I don't know if you are using d_d. Even D = 0 in the continuous state space is not zero in the discrete version.

y(k) = Cx(k) + Du(k)  

Here my implementation in the Autoware Control Toolbox.

 auto const &&mat1_ATs = I - A_ * Ts / 2.;
  auto const &&inv1_ATs = mat1_ATs.inverse();

  Ad_ = inv1_ATs * (I + A_ * Ts / 2.);
  Bd_ = inv1_ATs * B_ * Ts;
  Cd_ = C_ * inv1_ATs;
  Dd_ = D_ + C_ * Bd_ / 2.;

Dd = D + C  *inv(1 - A*dt/2) *B*dt/2

Reply:
Thank you for the comment. I misunderstood. Since we can assume the system has a ZOH input, we don't need to convert the measurement equation.